File: 1386-cinema-seat-allocation/1386. Cinema Seat Allocation.cpp
class Solution {
public:

    int calculateMaxSeats(vector<int> &bs)
    {
        //check for seat 2
        auto it = lower_bound(bs.begin(), bs.end(), 2);
        int nextSeat = 0;
        if(it != bs.end())nextSeat = *it;

        if(nextSeat > 9 || nextSeat == 0)return 2;
        else if(nextSeat > 5)return 1;

        else if(nextSeat <= 5)
        {
            //check for seat 4
            it = lower_bound(bs.begin(), bs.end(), 4);
            nextSeat = 0;
            if(it != bs.end())nextSeat = *it;

            if(nextSeat > 7 || nextSeat == 0)return 1;

            else if(nextSeat <= 5)
            {
                //check for seat 6
                it = lower_bound(bs.begin(), bs.end(), 6);
                nextSeat = 0;
                if(it != bs.end())nextSeat = *it;

                if(nextSeat > 9 || nextSeat == 0)return 1;
            }
        }

        return 0;
    } 

    int maxNumberOfFamilies(int n, vector<vector<int>>& r) {

        int ans = 0;
        map<int, vector<int>> mpp;
        sort(r.begin(), r.end());

        for(auto it : r)mpp[it[0]].push_back(it[1]);

        for(auto it : mpp)ans += calculateMaxSeats(it.second);

        ans += (n - mpp.size())*2;

        return ans;
    }
};

File: 1386-cinema-seat-allocation/1386. Cinema Seat Allocation.java
class Solution {
    public int maxNumberOfFamilies(int k, int[][] arr) {
        Arrays.sort(arr, (a, b) -> Integer.compare(a[0], b[0]));
        HashMap<Integer, HashSet<Integer>> mpp = new HashMap();
        int ans = k * 2;
        int n = arr.length;

        for (int i = 0; i < n; i++) {
            if (arr[i][1] >= 2 && arr[i][1] <= 9) {
                if (mpp.containsKey(arr[i][0]))
                    mpp.get(arr[i][0]).add(arr[i][1]);
                else {
                    mpp.put(arr[i][0], new HashSet<>());
                    mpp.get(arr[i][0]).add(arr[i][1]);
                }
            }
        }

        for (int it : mpp.keySet()) {
            ans -= 2;
            HashSet<Integer> st = mpp.get(it);

            //check left
            if (!st.contains(2) && !st.contains(3) && !st.contains(4) && !st.contains(5)) {
                ans++;
                continue;
            }

            // check middle

            if (!st.contains(4) && !st.contains(5) && !st.contains(6) && !st.contains(7)) {
                ans++;
                continue;
            }

            //check right
            if (!st.contains(6) && !st.contains(7) && !st.contains(8) && !st.contains(9)) {
                ans++;
                continue;
            }

            // IO.println(ans + " " + it);
        }

        return ans;
    }
}